cmd/tools/gendynamicconfig/main.go

```diff
@@ -0,0 +1,266 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/parser"
+	"go/token"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"text/template"
+)
+
+type (
+	settingType struct {
+		Name   string
+		GoType string
+		Index  int
+	}
+	settingPrecedence struct {
+		Name       string
+		GoArgs     string
+		GoArgNames string
+		Index      int
+	}
+)
+
+var (
+	types = []*settingType{
+		{
+			Name:   "Bool",
+			GoType: "bool",
+		},
+		{
+			Name:   "Int",
+			GoType: "int",
+		},
+		{
+			Name:   "Float",
+			GoType: "float64",
+		},
+		{
+			Name:   "String",
+			GoType: "string",
+		},
+		{
+			Name:   "Duration",
+			GoType: "time.Duration",
+		},
+		{
+			Name:   "Map",
+			GoType: "map[string]any",
+		},
+	}
+	precedences = []*settingPrecedence{
+		{
+			Name:       "Global",
+			GoArgs:     "()",
```

Comment:
Nit: we should add the parens in with the template. We shouldn't have to specify them for every precedence